Netable/Netable/Netable.swift

```diff
@@ -73,82 +80,93 @@ open class Netable {
             if T.Parameters.self != Empty.self {
                 try urlRequest.encodeParameters(for: request)
             }
-        } catch let error as NetableError {
-            logDestination.log(event: .requestFailed(error: error))
-            completion(.failure(error))
-            return RequestIdentifier(id: 0, session: self)
         } catch {
-            let unknownError = NetableError.unknownError(error)
-            logDestination.log(event: .requestFailed(error: unknownError))
-            completion(.failure(unknownError))
-            return RequestIdentifier(id: 0, session: self)
+            let netableError = (error as? NetableError) ?? NetableError.unknownError(error)
+            logDestination.log(event: .requestCreationFailed(urlString: request.path, error: netableError))
+            completion(.failure(netableError))
+            return RequestIdentifier(id: "invalid", session: self)
         }
 
         headers.forEach { key, value in
             urlRequest.setValue(value, forHTTPHeaderField: key)
         }
 
+        return startRequestTask(request, urlRequest: urlRequest, id: UUID().uuidString, retriesLeft: retryConfiguration.count, completion: completion)
+    }
+
+    private func startRequestTask<T: Request>(_ request: T, urlRequest: URLRequest, id: String, retriesLeft: UInt, completion: @escaping (Result<T.FinalResource, NetableError>) -> Void) -> RequestIdentifier {
         let startTimestamp = CACurrentMediaTime()
-        logDestination.log(event: .requestStarted(
+
+        let requestInfo = LogEvent.RequestInfo(
             urlString:  urlRequest.url?.absoluteString ?? "UNDEFINED",
             method: request.method,
             headers: urlRequest.allHTTPHeaderFields ?? [:],
             params: try? request.parameters.toParameterDictionary(encodingStrategy: request.jsonKeyEncodingStrategy))
-        )
 
-        let task = urlSession.dataTask(with: urlRequest) { data, response, error in
-            defer {
-                let endTimestamp = CACurrentMediaTime()
-
-                let userInfo = NetableNotification.userInfo(
-                    forRequest: urlRequest,
-                    data: data,
-                    response: response,
-                    duration: endTimestamp - startTimestamp,
-                    error: error
-                )
-
-                NotificationCenter.default.post(
-                    name: Notification.Name.NetableRequestDidComplete,
-                    object: self,
-                    userInfo: userInfo
-                )
+        logDestination.log(event: .requestStarted(request: requestInfo))
```

Comment:
The changelog makes it seem like _all_ log events are now run on the main thread, but isn't it possible log events like this aren't?

Why is it important log events are run on the main thread? If it is important, is it maybe worth baking into `logDestination.log`?

Reply:
Yeah, I guess it's only running everything on the main thread IF all of the incoming calls are also happening on the main thread, which in practice they probably always are, but we shouldn't be depending on.

Main reason that I think they should all be on the main thread is because of the idea that stuff that was done with notifications could / should be done with logs instead. Once the logging is potentially doing something a little more complicated than just formatting the text and passing on to another (presumably) thread safe print function, it's a little risky to just have those coming back on whatever random thread.

The notification already had this problem (any code that used it needed to dispatch to the main thread), it seems like a good idea to just do that by default. Though maybe we need a little more sophisticated approach (always dispatch to main, even for calls we think are probably already on main, specify a user defined queue, etc.).

I don't know that it needs to be fixed to merge this, since I think the current behaviour is an improvement in practice, but we should probably file a bug to clean this up further.